File: src/ghsingo/internal/audio/tonal_bed.go
package audio

// TonalBedVoice plays a long-form tonal sample as a continuously looped
// L0.5 bed between the synth drone (L0) and the noise bed (L1). Unlike
// the legacy BGM path it is not a "background music track" — it is one
// of three always-on ambient layers sharing the same space and master.
//
// The voice loops with a short crossfade (default 50 ms) to hide the
// loop boundary and applies a single-pole lowpass plus an exponentially
// smoothed gain so #31's MixerV2 can modulate the bed from composer
// state without clicks.
type TonalBedVoice struct {
	pcm        []float32
	pos        int
	crossfade  int
	lp         float32
	a          float32
	gain       float32
	targetGain float32
	gainSlew   float32
}

// NewTonalBedVoice builds a TonalBedVoice from mono PCM. Pass an empty
// slice or nil to disable the layer; in that case NextSample returns 0.
func NewTonalBedVoice(sampleRate int, pcm []float32) *TonalBedVoice {
	t := &TonalBedVoice{
		pcm:       pcm,
		crossfade: sampleRate / 20,         // 50 ms
		a:         0.10,                    // ~1.5 kHz lowpass
		gainSlew:  1.0 / float32(sampleRate) * 4.0,
	}
	if t.crossfade > len(pcm)/2 {
		t.crossfade = len(pcm) / 2
	}
	return t
}

// SetTargetGain queues a new gain target (smoothed over ~250 ms).
func (t *TonalBedVoice) SetTargetGain(g float32) {
	if g < 0 {
		g = 0
	}
	if g > 1 {
		g = 1
	}
	t.targetGain = g
}

// NextSample advances one sample and returns the lowpassed bed output.
// Returns 0 when the underlying PCM is empty.
func (t *TonalBedVoice) NextSample() float32 {
	if len(t.pcm) == 0 {
		return 0
	}
	loopLen := len(t.pcm)
	pos := t.pos % loopLen
	s := t.pcm[pos]

	if t.crossfade > 0 && t.crossfade < loopLen/2 {
		tail := loopLen - t.crossfade
		if pos >= tail {
			fadeOut := float32(loopLen-pos) / float32(t.crossfade)
			fadeIn := 1.0 - fadeOut
			headPos := pos - tail
			s = s*fadeOut + t.pcm[headPos]*fadeIn
		}
	}
	t.pos++

	t.lp += t.a * (s - t.lp)

	if t.gain < t.targetGain {
		t.gain += t.gainSlew
		if t.gain > t.targetGain {
			t.gain = t.targetGain
		}
	} else if t.gain > t.targetGain {
		t.gain -= t.gainSlew
		if t.gain < t.targetGain {
			t.gain = t.targetGain
		}
	}

	return t.lp * t.gain
}

File: src/rsghsing/src/audio/tonal_bed.rs
//! L0.5 sample-based tonal bed. Ported from `internal/audio/tonal_bed.go`.

pub struct TonalBedVoice {
    pcm: Vec<f32>,
    pos: usize,
    crossfade: usize,
    lp: f32,
    a: f32,
    gain: f32,
    target_gain: f32,
    gain_slew: f32,
}

impl TonalBedVoice {
    pub fn new(sample_rate: i32, pcm: Vec<f32>) -> Self {
        let mut crossfade = sample_rate as usize / 20; // 50 ms
        if crossfade > pcm.len() / 2 {
            crossfade = pcm.len() / 2;
        }
        TonalBedVoice {
            pcm,
            pos: 0,
            crossfade,
            lp: 0.0,
            a: 0.10, // ~1.5 kHz lowpass
            gain: 0.0,
            target_gain: 0.0,
            gain_slew: 1.0 / sample_rate as f32 * 4.0,
        }
    }

    pub fn set_target_gain(&mut self, g: f32) {
        self.target_gain = g.clamp(0.0, 1.0);
    }

    pub fn next_sample(&mut self) -> f32 {
        if self.pcm.is_empty() {
            return 0.0;
        }
        let loop_len = self.pcm.len();
        let pos = self.pos % loop_len;
        let mut s = self.pcm[pos];

        if self.crossfade > 0 && self.crossfade < loop_len / 2 {
            let tail = loop_len - self.crossfade;
            if pos >= tail {
                let fade_out = (loop_len - pos) as f32 / self.crossfade as f32;
                let fade_in = 1.0 - fade_out;
                let head_pos = pos - tail;
                s = s * fade_out + self.pcm[head_pos] * fade_in;
            }
        }
        self.pos += 1;

        self.lp += self.a * (s - self.lp);

        if self.gain < self.target_gain {
            self.gain += self.gain_slew;
            if self.gain > self.target_gain {
                self.gain = self.target_gain;
            }
        } else if self.gain > self.target_gain {
            self.gain -= self.gain_slew;
            if self.gain < self.target_gain {
                self.gain = self.target_gain;
            }
        }

        self.lp * self.gain
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_pcm_is_silent() {
        let mut t = TonalBedVoice::new(44_100, Vec::new());
        assert_eq!(t.next_sample(), 0.0);
    }

    #[test]
    fn crossfade_region_stays_continuous() {
        let pcm: Vec<f32> = (0..1_000).map(|i| (i as f32 / 1_000.0) - 0.5).collect();
        let mut t = TonalBedVoice::new(44_100, pcm);
        t.set_target_gain(1.0);
        let out: Vec<f32> = (0..3_000).map(|_| t.next_sample()).collect();
        for w in out.windows(2) {
            assert!((w[1] - w[0]).abs() < 0.05, "jump {} -> {}", w[0], w[1]);
        }
    }
}
